channels/rdpgfx/server/rdpgfx_main.c

```diff
@@ -0,0 +1,1239 @@
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_send_map_surface_to_window_pdu(RdpgfxServerContext* context, RDPGFX_MAP_SURFACE_TO_WINDOW_PDU* pdu)
+{
+	wStream* s = rdpgfx_server_packet_new(
+		     RDPGFX_CMDID_MAPSURFACETOWINDOW, 
+		     sizeof(RDPGFX_MAP_SURFACE_TO_WINDOW_PDU));
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "rdpgfx_server_packet_new failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	Stream_Write_UINT16(s, pdu->surfaceId); /* surfaceId (2 bytes) */
+	Stream_Write_UINT64(s, pdu->windowId); /* windowId (8 bytes) */
+	Stream_Write_UINT32(s, pdu->mappedWidth); /* mappedWidth (4 bytes) */
+	Stream_Write_UINT32(s, pdu->mappedHeight); /* mappedHeight (4 bytes) */
+
+	return rdpgfx_server_packet_send(context, s);
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_recv_frame_acknowledge_pdu(RdpgfxServerContext* context, wStream* s)
+{
+	RDPGFX_FRAME_ACKNOWLEDGE_PDU pdu;
+	UINT error = CHANNEL_RC_OK;
+
+	Stream_Read_UINT32(s, pdu.queueDepth); /* queueDepth (4 bytes) */
+	Stream_Read_UINT32(s, pdu.frameId); /* frameId (4 bytes) */
+	Stream_Read_UINT32(s, pdu.totalFramesDecoded); /* totalFramesDecoded (4 bytes) */
+
+	if (context)
+	{
+		IFCALLRET(context->FrameAcknowledge, error, context, &pdu);
+		if (error)
+			WLog_ERR(TAG, "context->FrameAcknowledge failed with error %lu", error);
+	}
+
+	return error;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_recv_cache_import_offer_pdu(RdpgfxServerContext* context, wStream* s)
+{
+	UINT16 index;
+	RDPGFX_CACHE_IMPORT_OFFER_PDU pdu;
+	RDPGFX_CACHE_ENTRY_METADATA* cacheEntries;
+	UINT error = CHANNEL_RC_OK;
+
+	Stream_Read_UINT16(s, pdu.cacheEntriesCount); /* cacheEntriesCount (2 bytes) */
+	pdu.cacheEntries = (RDPGFX_CACHE_ENTRY_METADATA*) 
+						calloc(pdu.cacheEntriesCount, sizeof(RDPGFX_CACHE_ENTRY_METADATA));
+
+	if (!pdu.cacheEntries)
+	{
+		WLog_ERR(TAG, "calloc failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	for (index = 0; index < pdu.cacheEntriesCount; index++)
+	{
+		cacheEntries = &(pdu.cacheEntries[index]);
+		Stream_Read_UINT64(s, cacheEntries->cacheKey); /* cacheKey (8 bytes) */
+		Stream_Read_UINT32(s, cacheEntries->bitmapLength); /* bitmapLength (4 bytes) */
+	}
+
+	if (context)
+	{
+		IFCALLRET(context->CacheImportOffer, error, context, &pdu);
+		if (error)
+			WLog_ERR(TAG, "context->CacheImportOffer failed with error %lu", error);
+	}
+
+	free(pdu.cacheEntries);
+
+	return error;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_recv_caps_advertise_pdu(RdpgfxServerContext* context, wStream* s)
+{
+	UINT16 index;
+	RDPGFX_CAPSET* capsSet;
+	RDPGFX_CAPSET capsSets[3];
+	RDPGFX_CAPS_ADVERTISE_PDU pdu;
+	UINT error = CHANNEL_RC_OK;
+
+	Stream_Read_UINT16(s, pdu.capsSetCount); /* capsSetCount (2 bytes) */
```

Comment:
In this function, you must add check so that you're sure that s has enough bytes to read.